internal/v5/resource.go

```
@@ -0,0 +1,176 @@
+	if err != nil {
+		return errgo.WithCausef(err, params.ErrNotFound, "")
+	}
+	ch, err := h.entityChannel(id)
+	if err != nil {
+		return errgo.Mask(err, errgo.Is(params.ErrNotFound))
+	}
+	r, err := h.Store.ResolveResource(id, rid.Name, rid.Revision, ch)
+	if err != nil {
+		return errgo.Mask(err, errgo.Is(params.ErrNotFound))
+	}
+	blob, err := h.Store.OpenResourceBlob(r)
+	if err != nil {
+		return errgo.Notef(err, "cannot open resource blob")
+	}
+	defer blob.Close()
+	header := w.Header()
+	setArchiveCacheControl(w.Header(), h.isPublic(id))
+	header.Set(params.ContentHashHeader, blob.Hash)
+
+	// TODO(rog) should we set connection=close here?
+	// See https://codereview.appspot.com/5958045
+	serveContent(w, req, blob.Size, blob)
+	return nil
+}
+
+func (h *ReqHandler) serveUploadResource(id *router.ResolvedURL, w http.ResponseWriter, req *http.Request) error {
+	if id.URL.Series == "bundle" {
+		return errgo.WithCausef(nil, params.ErrForbidden, "cannot upload a resource to a bundle")
+	}
+	name := strings.TrimPrefix(req.URL.Path, "/")
+	if !validResourceName(name) {
+		return badRequestf(nil, "invalid resource name")
+	}
+	hash := req.Form.Get("hash")
+	if hash == "" {
+		return badRequestf(nil, "hash parameter not specified")
+	}
+	if req.ContentLength == -1 {
+		return badRequestf(nil, "Content-Length not specified")
+	}
+	e, err := h.Cache.Entity(&id.URL, charmstore.FieldSelector("charmmeta"))
+	if err != nil {
+		// Should never happen, as the entity will have been cached
+		// when the charm URL was resolved.
+		return errgo.Mask(err, errgo.Is(params.ErrNotFound))
+	}
+	r, ok := e.CharmMeta.Resources[name]
+	if !ok {
+		return errgo.WithCausef(nil, params.ErrForbidden, "resource %q not found in charm metadata", name)
+	}
+	if r.Type != resource.TypeFile {
+		return errgo.WithCausef(nil, params.ErrForbidden, "non-file resource types not supported")
+	}
+	if filename := req.Form.Get("filename"); filename != "" && path.Ext(filename) != "" {
```

What if path.Ext(filename) -- "" but charmExt != ""? SHould we return an error? Isn't that a file extension mismatch?